libcloudforensics/aws.py

```diff
@@ -0,0 +1,726 @@
+    new resources in.
+  """
+  def __init__(self, default_availability_zone):
+    self.default_availability_zone = default_availability_zone
+    # The region is given by the zone minus the last letter
+    # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions
+    # -availability-zones.html#using-regions-availability-zones-describe
+    self.default_region = self.default_availability_zone[:-1]
+
+  def ClientApi(self, service, region=None):
+    """Create an AWS client object.
+
+    Attributes:
+      service (str): The AWS service to use.
+      region (str): The region is which to create new resources in.
+
+    Returns:
+      boto3.Session.Client: An AWS EC2 client object.
+    """
+    if region:
+      return CreateSession().client(
+          service_name=service, region_name=region
+      )
+    return CreateSession().client(
+        service_name=service, region_name=self.default_region
+    )
+
+  def ResourceApi(self, service, region=None):
+    """Create an AWS resource object.
+
+    Attributes:
+      service (str): The AWS service to use.
+      region (str): The region is which to create new resources in.
+
+    Returns:
+      boto3.Session.Resource: An AWS EC2 resource object.
+    """
+    if region:
+      return CreateSession().resource(
+          service_name=service, region_name=region
+      )
+    return CreateSession().resource(
+        service_name=service, region_name=self.default_region
+    )
+
+  def ListInstances(self, region=None, filters=None):
+    """List instances of an AWS account.
+
+    Attributes:
+      region (str): The region from which to list instances.
+      filters (list(dict)): Filters for the query.
+
+    Returns:
+      dict: Dictionary with name and metadata for each instance.
+
+    Example usage:
+      ListInstances(region='us-east-1', filters=[
+        dict(Name='instance-id', Values=['a-particular-instance-id'])
+      ])
+    """
+    if not filters:
+      filters = []
+
+    instances = dict()
+    have_all_tokens = False
+    next_token = None
+
+    while not have_all_tokens:
+      if next_token:
+        response = self.ClientApi(
+            EC2_SERVICE, region=region).describe_instances(
+                Filters=filters,
+                NextToken=next_token)
+      else:
+        response = self.ClientApi(
+            EC2_SERVICE, region=region).describe_instances(
+                Filters=filters)
+
+      for reservation in response["Reservations"]:
+        for instance in reservation["Instances"]:
+          instance_name = None
+          if instance.get('Tags'):
+            for tag in instance['Tags']:
+              instance_name = tag.get('Value') if tag.get('Key') == 'Name' \
+                else None
+              if instance_name is not None:
+                break
+          # Terminated instances are filtered out
+          if instance['State']['Name'] != 'terminated':
```

Comment:
```python
if instance['State']['Name'] == 'terminated':
  continue
```

Remember, guard statements are your friends :) 

Also, I see you're very careful in the lines before this regarding the expected keys to be set, but not here. Is there a reason why?
If the code is never functional when keys are not present, it's OK to raise an exception that can be caught by the caller.

Reply:
If there's an instance in `reservation['Instances']`, then the EC2 API always include the state of that instance, among other informations (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instances). It's therefore expected that these keys are set if the API call was successful. However I realised that there was no try/except statement around this very API call. This has been added.

Reply:
Please add a comment saying that if the above exists, then these keys can be expected to be set.